File: AIReaderKindle/src/Domain/Search/BookSearch.cpp
#include "BookSearch.hpp"

#include <glib.h>

namespace BookSearch {

namespace {

std::vector<gunichar> lowered(const std::string& text) {
    std::vector<gunichar> characters;
    const char* p = text.c_str();
    const char* end = p + text.size();
    for (; p < end; p = g_utf8_next_char(p)) characters.push_back(g_unichar_tolower(g_utf8_get_char(p)));
    return characters;
}

/// True when the text at `p` reads as `query`, letter for letter, ignoring
/// case. `matchEnd` receives where the match stops.
bool matches(const char* p, const char* end, const std::vector<gunichar>& query, const char** matchEnd) {
    for (gunichar wanted : query) {
        if (p >= end || g_unichar_tolower(g_utf8_get_char(p)) != wanted) return false;
        p = g_utf8_next_char(p);
    }
    *matchEnd = p;
    return true;
}

bool endsSentence(gunichar c) {
    return c == '.' || c == '!' || c == '?' || c == 0x2026 /* … */;
}

/// Where the sentence holding `offset` begins: after a paragraph break, or
/// after the space that follows a full stop.
int sentenceStart(const std::string& text, int offset, int floor) {
    const char* base = text.c_str();
    const char* p = base + offset;
    while (p > base + floor) {
        const char* previous = g_utf8_prev_char(p);
        gunichar c = g_utf8_get_char(previous);
        if (c == '\n') return static_cast<int>(p - base);
        if (g_unichar_isspace(c) && previous > base && endsSentence(g_utf8_get_char(g_utf8_prev_char(previous)))) {
            return static_cast<int>(p - base);
        }
        p = previous;
    }
    return static_cast<int>(p - base);
}

/// One past where the sentence holding `offset` ends: at a paragraph break,
/// or a full stop followed by space or the end of the text.
int sentenceEnd(const std::string& text, int offset, int ceiling) {
    const char* base = text.c_str();
    const char* end = base + ceiling;
    for (const char* p = base + offset; p < end; p = g_utf8_next_char(p)) {
        gunichar c = g_utf8_get_char(p);
        if (c == '\n') return static_cast<int>(p - base);
        if (!endsSentence(c)) continue;
        const char* next = g_utf8_next_char(p);
        // Closing quotes and brackets belong to the sentence.
        while (next < end && (g_utf8_get_char(next) == 0xBB || g_utf8_get_char(next) == '"' || g_utf8_get_char(next) == ')'
                              || g_utf8_get_char(next) == 0x201D)) {
            next = g_utf8_next_char(next);
        }
        if (next >= end || g_unichar_isspace(g_utf8_get_char(next))) return static_cast<int>(next - base);
    }
    return ceiling;
}

/// Moves a cut point to a character boundary, then to the nearest space on
/// the far side, so a cut never splits a word.
int cutBackward(const std::string& text, int at, int floor) {
    const char* base = text.c_str();
    const char* p = base + at;
    while (p > base + floor && (static_cast<unsigned char>(*p) & 0xC0) == 0x80) --p;
    while (p > base + floor && !g_unichar_isspace(g_utf8_get_char(p))) p = g_utf8_prev_char(p);
    return static_cast<int>(p - base);
}

int cutForward(const std::string& text, int at, int ceiling) {
    const char* base = text.c_str();
    const char* p = base + at;
    while (p < base + ceiling && (static_cast<unsigned char>(*p) & 0xC0) == 0x80) ++p;
    while (p < base + ceiling && !g_unichar_isspace(g_utf8_get_char(p))) p = g_utf8_next_char(p);
    return static_cast<int>(p - base);
}

/// The stretch of text an excerpt covers, and whether either end was cut.
struct Cut {
    int from;
    int to;
    bool front;
    bool back;
};

Cut cut(const std::string& text, int start, int end, int reach) {
    int size = static_cast<int>(text.size());
    Cut cut;
    cut.from = sentenceStart(text, start, std::max(0, start - reach));
    cut.to = sentenceEnd(text, end, std::min(size, end + reach));
    cut.front = cut.from > 0 && text[cut.from - 1] != '\n' && start - cut.from >= reach;
    cut.back = cut.to < size && text[cut.to] != '\n' && cut.to - end >= reach;
    if (cut.front) cut.from = cutForward(text, cut.from, start);
    if (cut.back) cut.to = cutBackward(text, cut.to, end);
    return cut;
}

}  // namespace

std::vector<SearchHit> find(const std::string& text, const std::string& query, int chapter, int limit) {
    std::vector<SearchHit> hits;
    std::vector<gunichar> wanted = lowered(query);
    if (wanted.empty() || limit <= 0) return hits;

    const char* base = text.c_str();
    const char* end = base + text.size();
    const char* p = base;
    // Where the last excerpt ended: a second match in the same sentence is
    // the same passage, listed once.
    int covered = 0;
    while (p < end && static_cast<int>(hits.size()) < limit) {
        const char* matchEnd = nullptr;
        if (!matches(p, end, wanted, &matchEnd)) {
            p = g_utf8_next_char(p);
            continue;
        }
        int start = static_cast<int>(p - base);
        p = matchEnd;
        if (start < covered) continue;
        SearchHit hit = excerpt(text, start, static_cast<int>(matchEnd - base));
        hit.chapter = chapter;
        hits.push_back(hit);
        covered = cut(text, start, static_cast<int>(matchEnd - base), reach).to;
    }
    return hits;
}

SearchHit excerpt(const std::string& text, int start, int end, int reach) {
    Cut range = cut(text, start, end, reach);
    SearchHit hit;
    hit.offset = start;
    hit.excerpt = (range.front ? "…" : "") + text.substr(range.from, range.to - range.from) + (range.back ? "…" : "");
    int shift = range.front ? static_cast<int>(std::string("…").size()) : 0;
    hit.matchStart = start - range.from + shift;
    hit.matchEnd = end - range.from + shift;
    // Trim the whitespace a sentence boundary leaves behind.
    while (hit.matchStart > 0 && (hit.excerpt[0] == ' ' || hit.excerpt[0] == '\t')) {
        hit.excerpt.erase(0, 1);
        --hit.matchStart;
        --hit.matchEnd;
    }
    while (!hit.excerpt.empty() && hit.excerpt.back() == ' ') hit.excerpt.pop_back();
    return hit;
}

}  // namespace BookSearch

File: AIReader/AIReader/Domain/Search/BookSearch.swift
import Foundation

/// One place a search query occurs.
struct SearchHit: Equatable, Sendable, Identifiable {
    var bookID: Book.ID = 0
    var bookTitle = ""
    /// The reading-order item the match is in, for the caption.
    var chapter = 0
    /// UTF-16 offset of the match in the book's text.
    var offset = 0
    /// The sentence around the match, cut down when it runs long.
    var excerpt = ""
    /// Where the match sits inside `excerpt`, in UTF-16 units.
    var matchRange = 0..<0

    var id: String { "\(bookID):\(offset)" }

    var position: BookPosition { BookPosition(bookID: bookID, chapter: chapter, offset: offset) }
}

/// A place in a book: a reading-order item and a UTF-16 offset into the
/// book's text.
struct BookPosition: Equatable, Sendable {
    var bookID: Book.ID
    var chapter: Int
    var offset: Int
}

/// Finds a query in a book's text, case-insensitively, and cuts a readable
/// excerpt around every hit. Pure: no display, no disk.
enum BookSearch {
    /// How far an excerpt reaches on either side of its match, in UTF-16
    /// units, when the sentence runs longer.
    static let reach = 220

    /// Hits within `range` of `text`, in order, at most `limit` of them. A
    /// sentence with the query in it twice is one hit.
    static func find(in text: String, query: String, range: NSRange? = nil, limit: Int) -> [SearchHit] {
        let source = text as NSString
        let bounds = range ?? NSRange(location: 0, length: source.length)
        guard !query.isEmpty, limit > 0, bounds.length > 0 else { return [] }

        var hits: [SearchHit] = []
        var from = bounds.location
        // Where the last excerpt ended: a second match in the same sentence
        // is the same passage, listed once.
        var covered = 0
        while hits.count < limit, from < NSMaxRange(bounds) {
            let match = source.range(
                of: query,
                options: .caseInsensitive,
                range: NSRange(location: from, length: NSMaxRange(bounds) - from)
            )
            guard match.location != NSNotFound else { break }
            from = NSMaxRange(match)
            if match.location < covered { continue }
            hits.append(excerpt(in: source, match: match))
            covered = cut(source, match: match, reach: reach).to
        }
        return hits
    }

    /// The excerpt for a match: its sentence, or as much of it as fits within
    /// `reach` on either side.
    static func excerpt(in text: NSString, match: NSRange, reach: Int = reach) -> SearchHit {
        let cut = cut(text, match: match, reach: reach)
        var excerpt = text.substring(with: NSRange(location: cut.from, length: cut.to - cut.from))
        var start = match.location - cut.from
        var end = NSMaxRange(match) - cut.from
        // Trim the whitespace a sentence boundary leaves behind.
        while start > 0, excerpt.first == " " || excerpt.first == "\t" {
            excerpt.removeFirst()
            start -= 1
            end -= 1
        }
        while excerpt.last == " " { excerpt.removeLast() }
        if cut.front {
            excerpt = "…" + excerpt
            start += 1
            end += 1
        }
        if cut.back { excerpt += "…" }
        return SearchHit(offset: match.location, excerpt: excerpt, matchRange: start..<end)
    }

    // MARK: - Sentence boundaries

    /// The stretch of text an excerpt covers, and whether either end was cut.
    private struct Cut {
        var from: Int
        var to: Int
        var front: Bool
        var back: Bool
    }

    private static func cut(_ text: NSString, match: NSRange, reach: Int) -> Cut {
        let start = match.location
        let end = NSMaxRange(match)
        let size = text.length
        var cut = Cut(
            from: sentenceStart(text, at: start, floor: max(0, start - reach)),
            to: sentenceEnd(text, at: end, ceiling: min(size, end + reach)),
            front: false,
            back: false
        )
        cut.front = cut.from > 0 && text.character(at: cut.from - 1) != newline && start - cut.from >= reach
        cut.back = cut.to < size && text.character(at: cut.to) != newline && cut.to - end >= reach
        if cut.front { cut.from = cutForward(text, at: cut.from, ceiling: start) }
        if cut.back { cut.to = cutBackward(text, at: cut.to, floor: end) }
        return cut
    }

    private static let newline: unichar = 0x0A

    private static func endsSentence(_ c: unichar) -> Bool {
        c == 0x2E || c == 0x21 || c == 0x3F || c == 0x2026  // . ! ? …
    }

    private static func isSpace(_ c: unichar) -> Bool {
        UnicodeScalar(c).map { CharacterSet.whitespacesAndNewlines.contains($0) } ?? false
    }

    private static func closesSentence(_ c: unichar) -> Bool {
        c == 0xBB || c == 0x22 || c == 0x29 || c == 0x201D  // » " ) ”
    }

    /// Where the sentence holding `offset` begins: after a paragraph break, or
    /// after the space that follows a full stop.
    private static func sentenceStart(_ text: NSString, at offset: Int, floor: Int) -> Int {
        var p = offset
        while p > floor {
            let c = text.character(at: p - 1)
            if c == newline { return p }
            if isSpace(c), p - 1 > 0, endsSentence(text.character(at: p - 2)) { return p }
            p -= 1
        }
        return p
    }

    /// One past where the sentence holding `offset` ends: at a paragraph
    /// break, or a full stop followed by space or the end of the text.
    private static func sentenceEnd(_ text: NSString, at offset: Int, ceiling: Int) -> Int {
        var p = offset
        while p < ceiling {
            let c = text.character(at: p)
            if c == newline { return p }
            if endsSentence(c) {
                var next = p + 1
                // Closing quotes and brackets belong to the sentence.
                while next < ceiling, closesSentence(text.character(at: next)) { next += 1 }
                if next >= ceiling || isSpace(text.character(at: next)) { return next }
            }
            p += 1
        }
        return ceiling
    }

    /// Moves a cut point to the nearest space on the far side, so a cut never
    /// splits a word or a surrogate pair.
    private static func cutBackward(_ text: NSString, at: Int, floor: Int) -> Int {
        var p = at
        while p > floor, UTF16.isTrailSurrogate(text.character(at: p)) { p -= 1 }
        while p > floor, !isSpace(text.character(at: p)) { p -= 1 }
        return p
    }

    private static func cutForward(_ text: NSString, at: Int, ceiling: Int) -> Int {
        var p = at
        while p < ceiling, UTF16.isTrailSurrogate(text.character(at: p)) { p += 1 }
        while p < ceiling, !isSpace(text.character(at: p)) { p += 1 }
        return p
    }
}
